cmd/docker/docker.go

```diff
@@ -82,9 +99,53 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
 	cmd.AddCommand(newDaemonCommand())
 	commands.AddCommands(cmd, dockerCli)
 
+	// The Args is handled by ValidateArgs in cobra, which does not allows a pre-hook.
+	// As a result, here we replace the existing Args validation func to a wrapper,
+	// where the wrapper will check to see if the feature is supported or not.
+	// The Args validation error will only be returned if the feature is supported.
+	visitAll(cmd, func(ccmd *cobra.Command) {
+		// if there is no tags for a command or any of its parent,
+		// there is no need to wrap the Args validation.
+		if !hasTags(ccmd) {
+			return
+		}
+
+		cmdArgs := ccmd.Args
+		if cmdArgs != nil {
+			ccmd.Args = func(cmd *cobra.Command, args []string) error {
+				if dockerCli.Client() == nil { // when using --help, PersistenPreRun is not called, so initialization is needed.
+					// flags must be the top-level command flags, not cmd.Flags()
+					opts.Common.SetDefaultOptions(flags)
+					dockerPreRun(opts)
+					dockerCli.Initialize(opts)
+				}
+				if err := isSupported(cmd, dockerCli.Client().ClientVersion(), dockerCli.HasExperimental()); err != nil {
+					return err
+				}
+				return cmdArgs(cmd, args)
+			}
+		}
+	})
+
 	return cmd
 }
 
+// visitAll will traverse all commands from the root.
+// This is different from the VisitAll of cobra.Command where only parents
+// are checked.
+func visitAll(root *cobra.Command, fn func(*cobra.Command)) {
+	var traverse func(r *cobra.Command, fn func(*cobra.Command))
```

Comment:
Thanks @LK4D4 for the review. The PR has been updated and `visitAll` has been simplified. Please take a look and let me know if there are any issues.